src/runtime/rpc/minrpc/minrpc_server.h

```diff
@@ -0,0 +1,598 @@
+            return;
+          }
+          default: {
+            this->ThrowError(RPCServerStatus::kUnknownRPCCode);
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  void Shutdown() {
+    arena_.FreeAll();
+    io_.Close();
+  }
+
+  void HandleNormalCallFunc() {
+    uint64_t call_handle;
+    TVMValue* values;
+    int* tcodes;
+    int num_args;
+    TVMValue ret_value[3];
+    int ret_tcode[3];
+
+    this->Read(&call_handle);
+    RecvPackedSeq(&values, &tcodes, &num_args);
+
+    int call_ecode = TVMFuncCall(
+        reinterpret_cast<void*>(call_handle),
+        values, tcodes, num_args,
+        &(ret_value[1]), &(ret_tcode[1]));
+
+    if (call_ecode == 0) {
+      // Return value encoding as in LocalSession
+      int rv_tcode = ret_tcode[1];
+      ret_tcode[0] = kDLInt;
+      ret_value[0].v_int64 = rv_tcode;
+      if (rv_tcode == kTVMNDArrayHandle) {
+        ret_tcode[1] = kTVMDLTensorHandle;
+        ret_value[2].v_handle = ret_value[1].v_handle;
+        ret_tcode[2] = kTVMOpaqueHandle;
+        this->ReturnPackedSeq(ret_value, ret_tcode, 3);
+      } else if (rv_tcode == kTVMPackedFuncHandle ||
+                 rv_tcode == kTVMModuleHandle) {
+        ret_tcode[1] = kTVMOpaqueHandle;
+        this->ReturnPackedSeq(ret_value, ret_tcode, 2);
+      } else {
+        this->ReturnPackedSeq(ret_value, ret_tcode, 2);
+      }
+    } else {
+      this->ReturnLastTVMError();
+    }
+  }
+
+  void HandleCopyFromRemote() {
+    uint64_t handle, offset, num_bytes;
+    TVMContext ctx;
+    DLDataType type_hint;
+
+    this->Read(&handle);
+    this->Read(&offset);
+    this->Read(&num_bytes);
+    this->Read(&ctx);
+    this->Read(&type_hint);
+
+    char* data_ptr;
+    int call_ecode = 0;
+    if (ctx.device_type == kDLCPU) {
+      data_ptr = reinterpret_cast<char*>(handle) + offset;
+    } else {
+      data_ptr = this->ArenaAlloc<char>(num_bytes);
+      call_ecode = TVMDeviceCopyDataFromTo(
+              reinterpret_cast<void*>(handle), offset,
+              data_ptr, 0, num_bytes,
+              ctx, DLContext{kDLCPU, 0},
+              type_hint, nullptr);
+    }
+
+    if (call_ecode == 0) {
+      RPCCode code = RPCCode::kCopyAck;
+      uint64_t packet_nbytes = sizeof(code) + num_bytes;
+
+      this->Write(packet_nbytes);
+      this->Write(code);
+      this->WriteArray(data_ptr, num_bytes);
+    } else {
+      this->ReturnLastTVMError();
+    }
+  }
+
+  void HandleCopyToRemote() {
+    uint64_t handle, offset, num_bytes;
+    TVMContext ctx;
+    DLDataType type_hint;
+
+    this->Read(&handle);
+    this->Read(&offset);
+    this->Read(&num_bytes);
+    this->Read(&ctx);
+    this->Read(&type_hint);
+    int call_ecode = 0;
+
+    if (ctx.device_type == kDLCPU) {
+       char* dptr = reinterpret_cast<char*>(handle) + offset;
+       this->ReadArray(dptr, num_bytes);
+    } else {
+      char* temp_data = this->ArenaAlloc<char>(num_bytes);
+      this->ReadArray(temp_data, num_bytes);
```

Comment:
for the micro runtime, we will need to avoid this copy step, here and in CopyFromRemote. is there a corresponding change we need to make to TVMDevice API to facilitate that?

Reply:
It only happens in the non-CPU case(which need the copy).

Reply:
so for micro would we similarly override kDLMicroDev

Reply:
hmm, on device microTVM rpc server, I believe it is better to set the device code to directly be kDLCPU(as it is CPU from the pt of view of the microcontroller).

If the code is kDLMicroDev, that means the current machine is a host machine that drives the micro tvm, and the temp memory is on the host